Sprint 2

MikhailSamigullin opened the pull request.

MikhailSamigullin committed:
Add style tools

Add ESlint, Stylelint to project.
Auto fix scss files with Stylelint.
Disaible in "rules" - "selector-class-pattern" becase it's working bad
with BEM naming of selectors.
Colors is fixed to kebab view.

Add devDependencies in package.json.

MikhailSamigullin committed:
Go from partials to components

This is first working app with routing.
Here's a lot of mess in files and files structure. Typing is bad.
Problem with styles.

To do fix on this week.

MikhailSamigullin committed:
Create and update components and pages

Also install and setting Eslint, Lint and tsconfig

MikhailSamigullin committed:
Change server logic

Project migrate to SPA and it's cause to change app.js with server logic

Correct package.json:
delete unused parts and scripts,
change start and dev scripts.

MikhailSamigullin committed:
Merge branch 'main' into sprint_2

MikhailSamigullin committed:
Fix tsconfig.json and delete console.logs

MikhailSamigullin committed:
Merge branch 'sprint_2' of https://github.com/MikhailSamigullin/middle.messenger.praktikum.yandex into sprint_2

MikhailSamigullin committed:
Delete comments in tsconfig

MikhailSamigullin committed:
Fix tsconfig

MikhailSamigullin committed:
Fix tsconfig again

MikhailSamigullin committed:
Revive project from backup

Here was problem with branche and commits

MikhailSamigullin committed:
Change style for stylelint test

MikhailSamigullin committed:
Add stylelintignore

MikhailSamigullin committed:
Fix stylelint

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)


К сожалению, работу придется отклонить без потери итерации:

* нет файла `Readme`. Его нужно сделать по чек-листу: указать ссылку задеплоенное приложение, описание, макет в фигме и тд

MikhailSamigullin committed:
Upload README file

gennady-bars commented on package.json:
```diff
@@ -7,24 +7,40 @@
+    "@parcel/packager-raw-url": "2.7.0",
+    "@parcel/transformer-sass": "2.7.0",
+    "@parcel/transformer-webmanifest": "2.7.0",
+    "@typescript-eslint/eslint-plugin": "^5.43.0",
```
Лучше задавать **фиксированные версии** зависимостям. Если нечаянно удалить `package-lock.json` и попытаться повторно установить зависимости с помощью команды `npm install`, то **npm** установит последние версии зависимостей, что может привести к ошибкам.

MikhailSamigullin replied:
Когда устанавливал дополнительные пакеты, забыл убрать " ^ ". Исправил

gennady-bars commented on package.json:
```diff
@@ -7,24 +7,40 @@
   },
   "keywords": [],
   "author": "MikhailSamigullin",
   "license": "ISC",
```
нужно ограничить версию `node` в `package.json` примерно так: `"engines" : { "node" : ">=12" }`, чтобы она была не ниже `12` версии. Это есть в чек-листе

MikhailSamigullin replied:
используя engines не получилось решить возникшие проблемы.Создал файл .nvmrc  с текущей версией ноды

gennady-bars commented on src/components/BackButton/backButton.module.css:
```diff
@@ -0,0 +1,128 @@
+  background-size: 30%;
+  top: 30px;
+  left: 0;
+}
```
В проекте в некоторых файлах есть ошибка `EOF`. Для `git` важно наличие **пустой строки** в конце файла. Подробнее тут: https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log Можно в настройки IDE добавить `insert_final_newline = true`

MikhailSamigullin replied:
Изменил настройку в IDE. Добавил пустую строку где необходимо. В основном проблема в css файлах, которые компилируются из scss.

gennady-bars commented on src/components/Input/input.ts:
```diff
@@ -0,0 +1,58 @@
+export class Input extends Block {
+  constructor(props: InputProps) {
+    super('input', {
+      type: props.type,
+      name: props.name,
+      value: props.value,
+      text: props.text,
+      required: props.required,
```
не нужно расписывать все элементы пропс опять. Можно использовать `spread`-оператор


```js

    super('input', {
      ...props,
      style: styles,
      events: {
      focus: (e: FocusEvent) => {

```

gennady-bars commented on src/components/Input/input.ts:
```diff
@@ -0,0 +1,58 @@
+      required: props.required,
+      style: styles,
+      events: {
+      focus: (e: FocusEvent) => {
+        e.preventDefault();
+        const element = e.target as HTMLInputElement;
+        const elementName: string = element.name;
+        const valid: any =  validation(elementName, element.value);
+        inputFields.map((item) => {
+          if (elementName === item) {
+            const label = document.querySelector(`#${item}`) as HTMLElement;
+             label.textContent = valid.text;    
+          }
+        }) 
+      },
+      blur: (e: FocusEvent) => {
+        e.preventDefault();
+        const element = e.target as HTMLInputElement;
+        const valid: any =  validation(element.name, element.value);
+        inputFields.map((item) => {
+          if (element.name === item) {
+            const label = document.querySelector(`#${item}`) as HTMLElement;
+             label.textContent = valid.text; 
+          }
+        })
+      },
```
функция валидация везде одинаковая. Не нужно дублировать код


ее нужно вынести в отдельную функцию `validateInput` 



```js



export const validateInput = (e: FocusEvent) => {
  e.preventDefault();
  const element = e.target as HTMLInputElement;
  const elementName: string = element.name;
  const valid: any =  validation(elementName, element.value);
  inputFields.map((item) => {
    if (elementName === item) {
      const label = document.querySelector(`#${item}`) as HTMLElement;
       label.textContent = valid.text;    
    }
  }) 
}

```

и использовать в коде

```js

      events: {
        focus: validateInput,
        blur: validateInput,

```


и теперь будет аккуратный код

gennady-bars commented on src/components/Input/input.ts:
```diff
@@ -0,0 +1,58 @@
+          }
+        })
+      },
+      submit: (e: MouseEvent) => {
```
`submit` вешается на форму, а не на инпуты

gennady-bars commented on src/components/Button/button.ts:
```diff
@@ -0,0 +1,83 @@
+      type: props.type,
+      style: styles,
+      events: {
+      submit: (e: Event) => {
```
**Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита , так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. **Это нужно исправить везде, где есть инпуты и форма**

gennady-bars commented on src/utils/Block.ts:
```diff
@@ -0,0 +1,186 @@
+import { nanoid } from 'nanoid';
+import Handlebars from 'handlebars';
+
+export class Block {
```
**Можно лучше**

лучше всего передавать тип пропсов в виде дженерика в `Block`, чтобы можно было конкретный тип пропсов всегда иметь под каждый отдельный экземпляр класса `Block`. Будет очень универсально 

```js
abstract class Block<Props extends Record<string, any> = unknown> {

```

gennady-bars commented on src/utils/Block.ts:
```diff
@@ -0,0 +1,186 @@
+    FLOW_CDM: 'flow:component-did-mount',
+    FLOW_RENDER: 'flow:rendeAr',
+    FLOW_CDU: 'flow:component-did-update',
+  };  
```
тут нужно добавить `  } as const;  `, чтобы зафиксировать литеральный тип строк


```js

  } as const;  

```

Подробнее https://igorfonin.gitbook.io/typescript-book-ru/typescript-type-system/literals

gennady-bars commented on src/utils/Block.ts:
```diff
@@ -0,0 +1,186 @@
+  private _element : HTMLElement | null = null;
+  private _meta : {tagName: string, props: any};
+  public refs: Record<string, Block> = {};
+  protected props: Record<string, unknown>;
```
`Record<string, unknown>;`


у пропсов внутри `Block` должен быть единый тип. Не нужно его обявлять постоянно заново. Он должен приходить либо из дженерика (смотрите мой совет) или объявить снаружи 1 раз

gennady-bars commented on src/utils/Block.ts:
```diff
@@ -0,0 +1,186 @@
+    return {props, children};
+  }
+
+  _addEvent() {
```
В теории давался еще метод `_removeEvents`, который правильно удаляет обработчики перед размонтированием блока. Нужно его добавить в `Block`.

gennady-bars commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,73 @@
+export default class HTTPTransport {
+  
+
+  get = (url: string, options = {}) => this.request(url, {
```
**Можно лучше**
Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище

```jsx

// создаем тип метода
type HTTPMethod = (url: string, options?: Options) => Promise<unknown>

export class HTTPTransport {

  // используем тип и удаляем дублирование в аргументах
  get: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.GET}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  put: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.PUT}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  post: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.POST}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  delete: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.DELETE}, options.timeout)
  )

```

MikhailSamigullin replied:
Если я правильно, то и request тоже можно так же типизировать

gennady-bars commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,73 @@
+    xhr.open(
+      options.method,
+      options.method === METHODS.GET && !!options.data
+        ? `${url}${queryStringify(options.data)}`
```
`queryStringify` лучше сразу в методе `get` вызывать, чтобы не проверять `METHODS.GET` при каждом запросе

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная



* `Readme` хорошо оформлен
* Проект задеплоен и отлично работает
* Отлично, что не забыли про `.gitignore`
* Хорошая структура папок и файлов
* Добавлен `ESLint` и `Stylelint`
* Отлично, что цвета вынесены в `css`-константы
* Отлично, что не забываете про тег `main` для главной части страницы
* Хорошая валидация
* Отлично, что отображаете текст ошибки под инпутами

но есть некоторые недочеты: 

* нужно ограничить версию `node` в `package.json` примерно так: `"engines" : { "node" : ">=12" }`, чтобы она была не ниже `12` версии. Это есть в чек-листе 
* В проекте в некоторых файлах есть ошибка `EOF`. Для `git` важно наличие **пустой строки** в конце файла. Подробнее тут: https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log Можно в настройки IDE добавить `insert_final_newline = true`
* `submit` вешается на форму, а не на инпуты
* **Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. **Это нужно исправить везде, где есть инпуты и форма**
* у пропсов внутри `Block` должен быть единый тип. Не нужно его обявлять постоянно заново. Он должен приходить либо из дженерика (смотрите мой совет) или объявить снаружи 1 раз
* В теории давался еще метод `_removeEvents`, который правильно удаляет обработчики перед размонтированием блока. Нужно его добавить в `Block`.



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

MikhailSamigullin commented on src/pages/index.html:
```diff
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
```
Вернул favicon обратно

MikhailSamigullin committed:
Fix after review

Change:
- Typing Block and EventBus
- Add new line in the end of generated css files
- Change target for submit form
- Minor changes

gennady-bars commented on package.json:
```diff
@@ -7,24 +7,40 @@
+    "eslintfix": "eslint \"**/*.ts\" --fix",
+    "lintfix": "stylelint \"**/*.scss\" --fix",
+    "lint": "stylelint \"**/*.scss\"",
+    "dev": "parcel src/pages/index.html --no-cache",
```
Чтобы **автоматически запускался** проект в браузере, нужно дописать в скрипт `--open`

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
